segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/MetadataCleaner.java

```diff
@@ -179,6 +177,27 @@ protected void doStop() {
                 }, this.executor);
     }
 
+    private CompletableFuture<Void> deleteUnusedTransientSegments(long lastSeqNo) {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, this.traceObjectId, "deleteUnusedTransientSegments", lastSeqNo);
+        // Serialize only those segments that are still alive (not deleted or merged - those will get removed anyway).
+        val deletionTasks = this.metadata.getEvictionCandidates(lastSeqNo, this.config.getMaxConcurrentSegmentEvictionCount())
+                .stream()
+                .filter(sm -> sm.getType().isTransientSegment())
+                .filter(sm -> sm.getAttributes().get(Attributes.CREATION_EPOCH) < metadata.getContainerEpoch())
+                .map(sm -> metadataStore.deleteSegment(sm.getName(), this.config.getTransientSegmentDeleteTimeout()))
+                .collect(Collectors.toList());
+
+        return Futures.allOf(deletionTasks)
+                .thenRunAsync(() -> {
```

Comment:
Performance tip: `TRACE` is unlikely to be used at all, and guaranteed never to be used in production. This means that you'll be essentially creating a new callback and queuing it onto the executor every time you invoke this method even if it won't do anything. This has the potential of slowing down the system.

Consider doing this:
1. create a local variable `result` to hold the result of `Futures.allOf`
2. `if(log.isTraceEnabled())` then attach this callback to result (but do not include it in the return value. I.e., just do `result.thenRun`
3. Notice how I didn't even do `thenRunAsync`. Since it's such a small callback, no need to re-queue it on the executor; just run it inline. 
4. Return `result`.

For examples, search for `.isTraceEnabled` or `.isDebugEnabled` in the codebase; we've used this pattern in other places too.